hugr-core/src/hugr.rs

```diff
@@ -304,4 +303,158 @@ mod test {
+        const XA: ExtensionId = "EXT_A";
+        const XB: ExtensionId = "EXT_B";
+    }
+
+    #[rstest]
+    #[case([], XA.into())]
+    #[case([XA], XA.into())]
+    #[case([XB], ExtensionSet::from_iter([XA, XB]))]
+
+    fn infer_single_delta(
+        #[case] parent: impl IntoIterator<Item = ExtensionId>,
+        #[values(true, false)] remove: bool, // makes no difference when inferring
+        #[case] result: ExtensionSet,
+    ) {
+        let parent = ExtensionSet::from_iter(parent).union(TO_BE_INFERRED.into());
+        let (mut h, _) = build_ext_cfg(parent);
+        h.infer_extensions(remove).unwrap();
+        assert_eq!(h, build_ext_cfg(result).0);
+    }
+
+    #[test]
+    fn infer_removes_from_delta() {
+        let parent = ExtensionSet::from_iter([XA, XB]);
+        let mut h = build_ext_cfg(parent.clone()).0;
+        let backup = h.clone();
+        h.infer_extensions(false).unwrap();
+        assert_eq!(h, backup); // did nothing
+        h.infer_extensions(true).unwrap();
+        assert_eq!(h, build_ext_cfg(XA.into()).0);
+    }
+
+    #[test]
+    fn infer_bad_remove() {
+        let (mut h, mid) = build_ext_cfg(XB.into());
+        let backup = h.clone();
+        h.infer_extensions(false).unwrap();
+        assert_eq!(h, backup); // did nothing
+        let val_res = h.validate(&EMPTY_REG);
+        let expected_err = ExtensionError {
+            parent: h.root(),
+            parent_extensions: XB.into(),
+            child: mid,
+            child_extensions: XA.into(),
+        };
+        #[cfg(feature = "extension_inference")]
+        assert_eq!(
+            val_res,
+            Err(ValidationError::ExtensionError(expected_err.clone()))
+        );
+        #[cfg(not(feature = "extension_inference"))]
+        assert!(val_res.is_ok());
+
+        let inf_res = h.infer_extensions(true);
+        assert_eq!(inf_res, Err(expected_err));
+    }
+
+    fn build_ext_cfg(parent: ExtensionSet) -> (Hugr, Node) {
+        let ty = Type::new_function(FunctionType::new_endo(type_row![]));
+        let mut h = Hugr::new(ops::DFG {
+            signature: FunctionType::new_endo(ty.clone()).with_extension_delta(parent.clone()),
+        });
+        let root = h.root();
+        let mid = add_inliftout(&mut h, root, ty);
+        (h, mid)
+    }
+
+    fn add_inliftout(h: &mut Hugr, p: Node, ty: Type) -> Node {
+        let inp = h.add_node_with_parent(
+            p,
+            ops::Input {
+                types: ty.clone().into(),
+            },
+        );
+        let out = h.add_node_with_parent(
+            p,
+            ops::Output {
+                types: ty.clone().into(),
+            },
+        );
+        let mid = h.add_node_with_parent(
+            p,
+            ops::Lift {
+                type_row: ty.into(),
+                new_extension: XA,
+            },
+        );
+        h.connect(inp, 0, mid, 0);
+        h.connect(mid, 0, out, 0);
+        mid
+    }
+
+    #[rstest]
```

Comment:
i think these cases would benefit from comments next to each saying briefly why the result is expected

Reply:
Fair, done